cluster/gce/master.sh

```diff
@@ -0,0 +1,138 @@
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/../..
+source "${KUBE_ROOT}/cluster/gce/common.sh"
+
+# Provision master machine and configures network to give access to it over HTTPS.
+function master::provision {
+  master::configure-network
+  master::create
+
+  # Wait for last batch of jobs
+  wait-for-jobs
+}
+
+# Configures network to give access to master machine over HTTPS.
+function master::configure-network {
+  detect-project
+
+  echo "Creating firewall rule for HTTPS traffic to ${MASTER_NAME}"
+  gcloud compute firewall-rules create "${MASTER_NAME}-https" \
+    --project "${PROJECT}" \
+    --network "${NETWORK}" \
+    --target-tags "${MASTER_TAG}" \
+    --allow tcp:443 &
+}
+
+# Creates master machine with a static IP and persistent disk.
+function master::create {
+  KUBE_MASTER=${MASTER_NAME}
+  master::create-disk
+  master::create-ip
+  master::create-instance "${KUBE_MASTER_IP}" &
+}
+
+# Creates persistent disk for master machine.
+function master::create-disk {
+  detect-project
+
+  # We have to make sure the disk is created before creating the master VM, so
+  # run this in the foreground.
+  gcloud compute disks create "${MASTER_NAME}-pd" \
+    --project "${PROJECT}" \
+    --zone "${ZONE}" \
+    --type "${MASTER_DISK_TYPE}" \
+    --size "${MASTER_DISK_SIZE}"  
+}
+
+# Creates static IP address for master machine.
+function master::create-ip {
+  detect-project
+
+  # Reserve the master's IP so that it can later be transferred to another VM
+  # without disrupting the kubelets. IPs are associated with regions, not zones,
+  # so extract the region name, which is the same as the zone but with the final
+  # dash and characters trailing the dash removed.
+  local REGION=${ZONE%-*}
+
+  local attempt=0
+  while true; do
+    if ! gcloud compute addresses create "${MASTER_NAME}-ip" \
+      --project "${PROJECT}" \
+      --region "${REGION}" -q > /dev/null; then
+      if (( attempt > 4 )); then
+        echo -e "${color_red}Failed to create static ip $1 ${color_norm}" >&2
+        exit 2
+      fi
+      attempt=$(($attempt+1)) 
+      echo -e "${color_yellow}Attempt $attempt failed to create static ip $1. Retrying.${color_norm}" >&2
+      sleep $(($attempt * 5))
+    else
+      break
+    fi
+  done
+
+  KUBE_MASTER_IP=$(gcloud compute addresses describe "${MASTER_NAME}-ip" \
+    --project "${PROJECT}" \
+    --region "${REGION}" -q --format yaml | awk '/^address:/ { print $2 }')
+}
+
+# Creates the master instance. If called with an argument, the argument is
+# used as the name to a reserved IP address for the master. (In the case of
+# upgrade/repair, we re-use the same IP.)
+#
+# It requires a whole slew of assumed variables, partially due to to
+# the call to write-master-env. Listing them would be rather
+# futile. Instead, we list the required calls to ensure any additional
+# variables are set:
+#   ensure-temp-dir
+#   detect-project
+#   get-bearer-token
+#
+function master::create-instance {
+  detect-project
```

Comment:
The comment says that `detect-project` should have been called prior to calling this function.